src/riscv/lib/src/state_backend/proof_backend/proof.rs

```diff
@@ -317,110 +309,40 @@ fn deserialise_final_hash_encoding(
-                Tree::Leaf(size) => Ok(ModifyResult::LeafStop((size, leaf_tag))),
-            },
-            Tag::Node => match subtree {
-                Tree::Node(children) => Ok(ModifyResult::NodeContinue((), children)),
-                Tree::Leaf(_) => Err(DeserialiseError::UnexpectedNode),
-            },
-        }
-    })
-}
-
-fn deserialise_merkle_proof_values(
-    bytes: &mut impl Iterator<Item = u8>,
-    shape: Tree<(usize, LeafTag)>,
-) -> Result<MerkleProof, DeserialiseError> {
-    shape.map(&mut |(size, tag)| {
-        Ok(match tag {
-            LeafTag::Blind => MerkleProofLeaf::Blind(deserialise_final_hash_encoding(bytes)?),
-            LeafTag::Read => {
-                let mut data = Vec::with_capacity(size);
-                for _ in 0..size {
-                    data.push(bytes.next().ok_or(DeserialiseError::NotEnoughBytes)?);
-                }
-                MerkleProofLeaf::Read(data)
-            }
-        })
-    })
-}
-
-fn deserialise_merkle_proof(
-    bytes: &mut impl Iterator<Item = u8>,
-    full_shape: Shape,
-) -> Result<MerkleProof, DeserialiseError> {
-    let proof_shape = deserialise_merkle_proof_tags(bytes, full_shape)?;
-    deserialise_merkle_proof_values(bytes, proof_shape)
-}
-
-/// Given a [`Iterator<Item = u8>`], parse the contents into a [`Proof`].
+/// Deserialise a [`Proof`] from an iterator of bytes.
 ///
-/// A [`Shape`] of the full Merkle tree has to be provided to account for the proof being
-/// only a subtree of the actual full Merkle tree.
-/// Knowing the initial shape statically helps improve size of the encoding by
-/// not needing to know arity of different subtrees / different leaf sizes.
-///
-/// Ideally, the shape should be a compile time type/argument.
-/// However, for ease of implementation currently it is given in the form of a [`Shape`] argument
-pub fn deserialise_proof(
-    mut bytes: impl Iterator<Item = u8>,
-    full_merkle_tree_shape: Shape,
-) -> Result<Proof, DeserialiseError> {
-    let final_hash = deserialise_final_hash_encoding(&mut bytes)?;
-    let merkle_proof = deserialise_merkle_proof(&mut bytes, full_merkle_tree_shape)?;
-
-    // we expect the iterator to be fully consumed when deserialising a proof
-    if bytes.next().is_some() {
-        return Err(DeserialiseError::TooManyBytes);
-    }
+/// Obtain a [`Proof`] and the associated [`NodePvm<Verifier>`] backend.
+pub fn deserialise_proof<I: Iterator<Item = u8>>(
+    bytes: I,
+) -> Result<(Proof, NodePvm<Verifier>), FromProofError> {
+    let mut bytes_iter = bytes;
+    let final_state_hash = deserialise_final_hash(&mut bytes_iter)?;
+
+    let (space, proof_tree) = deserialise_stream::deserialise::<NodePvmLayout>(
+        bytes_iter.collect::<Vec<u8>>().as_slice(),
+    )?;
+
+    let pvm = NodePvm::bind(space);
+
+    let merkle_tree = match proof_tree {
+        OwnedProofPart::Absent => return Err(FromProofError::AbsentProof),
+        OwnedProofPart::Present(tree) => tree,
+    };
```

Comment:
Error as early as possible.

```suggestion
    let merkle_tree = match proof_tree {
        OwnedProofPart::Absent => return Err(FromProofError::AbsentProof),
        OwnedProofPart::Present(tree) => tree,
    };

    let pvm = NodePvm::bind(space);
```

Reply:
Changed [here](https://github.com/tezos/riscv-pvm/compare/a4eb5267b35192ae1b276134037e26787ab5cda0..44c8bd1236f532f066585a0c60d0278e3a68fd87).